migtools/oadp-non-admin
# Added Estimated Queue Position in the Status field of the NAB object

## Why the changes were made


To fix #20. Included in this fix is also small fix for the #119.

## How to test the changes made



There were unit and functional tests written to cover newly written functions within `function.go`.
There was `estimatedQueuePosition` added to the tests of the reconcile to include that coverage. Note no new tests in the `nonadminbackup_controller_test.go` were added, because it was straight forward to add test coverage to already existing test cases as part of the status check.

1. Written tests, please run:
```shell
$ make test
[...]
	github.com/migtools/oadp-non-admin/api/v1alpha1		coverage: 0.0% of statements
?   	github.com/migtools/oadp-non-admin/internal/common/constant	[no test files]
	github.com/migtools/oadp-non-admin/cmd		coverage: 0.0% of statements
ok  	github.com/migtools/oadp-non-admin/internal/common/function	0.018s	coverage: 50.3% of statements
	github.com/migtools/oadp-non-admin/internal/handler		coverage: 0.0% of statements
	github.com/migtools/oadp-non-admin/test/utils		coverage: 0.0% of statements
	github.com/migtools/oadp-non-admin/internal/predicate		coverage: 0.0% of statements
ok  	github.com/migtools/oadp-non-admin/internal/controller	14.405s	coverage: 58.2% of statements
```

2. To test the functionality on the test cluster:

- Create deployment as usual with the OADP and NonAdminBackup controllers
- Create non-admin namespace
- One could run the following shell script to create 20 NAB objects by setting `$ export BACKUP_NAMESPACE=my-non-admin-namespace`:
```shell
#!/bin/bash

if [ -z "$BACKUP_NAMESPACE" ]; then
  echo "ERROR: BACKUP_NAMESPACE environment variable is not set."
  exit 1
fi

counter=1

# 20 NABs to be created
while [ $counter -le 20 ]; do
  object_name="anothertestbackup-${counter}"
  
  cat <<EOF | oc apply -f -
apiVersion: oadp.openshift.io/v1alpha1
kind: NonAdminBackup
metadata:
  name: ${object_name}
  namespace: ${BACKUP_NAMESPACE}
spec:
  backupSpec:
    snapshotVolumes: true
    snapshotMoveData: true
    defaultVolumesToFsBackup: true
EOF

  echo "Created NonAdminBackup: ${object_name} in namespace ${BACKUP_NAMESPACE}"

  ((counter++))

  # Wait for 1 second before creating the next object
  sleep 1
done
```
- Observe estimatedQueuePositon within NAB Objects, can be done with:
```shell
$ while true; do oc get nonadminbackup -n "$BACKUP_NAMESPACE" -o jsonpath='{range .items[*]}{.status.queueInfo}{"\t"}{.metadata.name}{"\n"}{end}'|sort; sleep 1; done
```
